src/cmd/sanity_test/sanity_tests.go

```diff
@@ -105,260 +126,328 @@ func checkEmailRestoration(
+		ItemsById(restoreFolderID).
+		Children().
+		Get(ctx, nil)
 	if err != nil {
-		fmt.Printf("Error getting child folder: %v\n", err)
-		os.Exit(1)
+		fatal(ctx, "getting child folder", err)
 	}
 
-	for _, restoreData := range restoreResponses.GetValue() {
-		restoreName := *restoreData.GetName()
-
-		if restoreData.GetFile() != nil {
-			if *restoreData.GetSize() != file[restoreName] {
-				fmt.Printf("Size of file %s is different in drive %d and restored file: %d ",
-					restoreName,
-					file[restoreName],
-					*restoreData.GetSize())
+	for _, item := range restored.GetValue() {
+		var (
+			itemID   = ptr.Val(item.GetId())
+			itemName = ptr.Val(item.GetName())
+			itemSize = ptr.Val(item.GetSize())
+		)
+
+		if item.GetFile() != nil {
+			if itemSize != file[itemName] {
+				fmt.Println("File size does not match:")
+				fmt.Println("-  expected:", file[itemName])
+				fmt.Println("-  actual:", itemSize)
+				fmt.Println("Item:", itemName, itemID)
 				os.Exit(1)
 			}
 
 			continue
 		}
 
-		itemBuilder := client.DrivesById(driveID).ItemsById(*restoreData.GetId())
+		if item.GetFolder() == nil {
+			continue
+		}
 
-		if restoreData.GetFolder() != nil {
-			permissionColl, err := itemBuilder.Permissions().Get(context.TODO(), nil)
-			if err != nil {
-				fmt.Printf("Error getting permission: %v\n", err)
-				os.Exit(1)
-			}
+		var (
+			expectItem = folderPermission[itemID]
+			results    = permissionsIn(ctx, client, driveID, itemID, nil)
+		)
 
-			userPermission := []string{}
+		for pid, result := range results {
+			expect := expectItem[pid]
 
-			for _, perm := range permissionColl.GetValue() {
-				userPermission = perm.GetRoles()
-			}
-
-			if !reflect.DeepEqual(folderPermission[restoreName], userPermission) {
-				fmt.Printf("Permission mismatch for %s.", restoreName)
+			if !slices.Equal(expect, result) {
+				fmt.Println("permissions are not equal")
+				fmt.Println("-  expected: ", expect)
+				fmt.Println("-  actual: ", result)
+				fmt.Println("Item:", itemName, itemID)
+				fmt.Println("Permission:", pid)
 				os.Exit(1)
 			}
 		}
 	}
 }
+
+// ---------------------------------------------------------------------------
+// Helpers
+// ---------------------------------------------------------------------------
+
+func fatal(ctx context.Context, msg string, err error) {
+	logger.CtxErr(ctx, err).Error("test failure: " + msg)
+	fmt.Println(msg+":", err)
+	os.Exit(1)
+}
+
+func permissionsIn(
+	ctx context.Context,
+	client *msgraphsdk.GraphServiceClient,
+	driveID, itemID string,
+	init map[string][]string,
+) map[string][]string {
+	result := map[string][]string{}
+
+	pcr, err := client.
+		DrivesById(driveID).
+		ItemsById(itemID).
+		Permissions().
+		Get(ctx, nil)
+	if err != nil {
+		fatal(ctx, "getting permission", err)
+	}
+
+	if len(init) > 0 {
+		maps.Copy(result, init)
+	}
+
+	// check if permission are correct on folder
+	for _, p := range pcr.GetValue() {
+		var (
+			pid   = ptr.Val(p.GetId())
```

Comment:
I don't know if we can assume that restore will re-use the same permission IDs. It seems like OneDrive will try to reuse them if a permission for the same stuff already exists, but I don't think there's a guarantee. It may be better to compare role+grantedToV2

I think the above also depends a bit on how we want the compare to work. Since both role and grantedToV2 are both slices there's a possibility of checking if we got the same permission sets back from a "structure" perspective (i.e. same roles/grantedTo pairs) or just from a more broad perspective of "all the grantedTo entries that had role X are the same no matter what other roles may have been assigned in the same permission object." The latter would essentially merge grantedTo values across permission objects so that everything could be compared by role

Reply:
Leaving this for a follow-up. @neha-Gupta1

Reply:
Will take care of this in coming PR.

Reply:
Will take care of this in coming PR.

Reply:
@ashmrtn - Checks that I will be doing for Onedrive in https://github.com/alcionai/corso/pull/2938
- if file- verify file size matches
- if folder - check permission(role+grantedToV2) on folder matches